### DISCUSSION TITLE
Cooldown - How to display the Time Left on a specific command's cooldown

### DISCUSSION THREAD

**Discussion by samjesus8** [DSharpPlus/DSharpPlus#1417] [Support] [answered]
# Summary
I have a method in my Bot.cs file which throws an exception when it detects an active cooldown and lets the user know they must wait till the cooldown timer is finished before they can use a command again. I'm having trouble finding a way to display the time left on the command. So far I can only let the user know that a cooldown is in effect with no indication of how long is left.

# Details
My aim is to improve this method so I am able to get the time left on a specific command's cooldown and display it in a Discord Embedded message like I have in the following method

```
        private async Task OnCommandError(CommandsNextExtension sender, CommandErrorEventArgs e)
        {
            if (e.Exception is ChecksFailedException) 
            {
                Console.Error.WriteLine("Error " + e.Exception); //Writes error to console

                var cooldownErrorMessage = new DiscordMessageBuilder() //Message to notify user
                    .AddEmbed(new DiscordEmbedBuilder()
                    .WithColor(DiscordColor.Azure)
                    .WithTitle("You must wait for the cooldown to end")
                    );

                await e.Context.Channel.SendMessageAsync(cooldownErrorMessage);
            }
        }
```
So all I'm looking for is a way to get the Cooldown of an executed command and if the Cooldown limit is exceeded, throw this exception and send this message with a timer showing how long is left on the cooldown as I have commands that have different cooldowns

**Comment by Giggitybyte** [comment, marked as answer]
Cast the exception to `ChecksFailedException` then check if any of the attributes returned by the `FailedChecks` property are of type `CooldownAttribute`. You'll then be able to cast that to `CooldownAttribute` and call `GetRemainingCooldown()` to get the time remaining.

**Reply by samjesus8** [reply]
Yep thanks that makes so much sense, it's very faint on how this is actually done so this helped so much

**Comment by samjesus8** [comment]
For anyone who needs the solution to this, I'm not gonna be a dickhead and I'm just going to post my version of adding a timer to a cooldown:

```
        private async Task OnCommandError(CommandsNextExtension sender, CommandErrorEventArgs e)
        {
            if (e.Exception is ChecksFailedException) //Checking if its a failed command 
            {
                var cast = (ChecksFailedException)e.Exception; //Converting the exception into type ChecksFailedException
                string timeLeftCooldown = ""; //Variable to store cooldown remaining time

                foreach (var check in cast.FailedChecks) //Checking for any Cooldown Related errors
                {
                    var cooldown = (CooldownAttribute)check; //Casting as a Cooldown Attribute
                    TimeSpan timeLeft = cooldown.GetRemainingCooldown(e.Context); //Getting remaining cooldown
                    timeLeftCooldown = timeLeft.ToString(@"hh\:mm\:ss");
                }

                Console.Error.WriteLine("Error " + e.Exception);

                var cooldownErrorMessage = new DiscordMessageBuilder() //Message to send when this exception is thrown
                    .AddEmbed(new DiscordEmbedBuilder()
                    .WithColor(DiscordColor.DarkRed)
                    .WithTitle("You must wait for the cooldown to end!!")
                    .WithDescription("Remaining Time (HH:MM:SS) -> ***" + timeLeftCooldown + "***")
                    );

                await e.Context.Channel.SendMessageAsync(cooldownErrorMessage);
            }
        }
```

You will then need to register this in your CommandsNext config as a CommandErrored because you are not actually adding a set of commands but this is more of a check
